scripts/init.sh

```diff
@@ -3,92 +3,111 @@
-    TARGET="aarch64-unknown-linux-musl"
-    ;;
-Linux:arm*)     
-    TARGET="armv7-unknown-linux-musleabihf"
-    ;;
-Linux:i686:iSH*)
-    TARGET="ish"
-    ;;
-Linux:i686*)     
-    TARGET="i686-unknown-linux-musl"
-    ;;
-Linux:i586*)     
-    TARGET="i586-unknown-linux-musl"
-    ;;
-Linux:mips*)     
-    TARGET="mips-unknown-linux-musl"
-    ;;
-FreeBSD:x86_64*)
-    TARGET="x86_64-unknown-freebsd"
-    ;;
-FreeBSD:amd64*)
-    TARGET="x86_64-unknown-freebsd"
-    ;;
-FreeBSD:i686*)
-    TARGET="i686-unknown-freebsd"
-    ;;
-Darwin:x86_64*)    
-    TARGET="x86_64-apple-darwin"
-    ;;
-WindowsNT:x86_64*)    
-    TARGET="x86_64-pc-windows-msvc.exe"
-    ;;
-WindowsNT:i686*)    
-    TARGET="i686-pc-windows-msvc.exe"
-    ;;
-*)          
-    echo "Unsupported system ($(uname -a))"
-    exit 1
-    ;;
-esac
+main() {
+    case "$(uname -s):$(uname -m):$(uname -v):$(uname -a)" in
+    Linux:x86_64*)     
+        TARGET="x86_64-unknown-linux-musl"
+        ;;
+    Linux:aarch64:*:*Android*)
+        TARGET="aarch64-linux-android"
+        ;;
+    Linux:arm64*|Linux:aarch64*)
+        TARGET="aarch64-unknown-linux-musl"
+        ;;
+    Linux:arm*)     
+        TARGET="armv7-unknown-linux-musleabihf"
+        ;;
+    Linux:i686:iSH*)
+        TARGET="ish"
+        ;;
+    Linux:i686*)     
+        TARGET="i686-unknown-linux-musl"
+        ;;
+    Linux:i586*)     
+        TARGET="i586-unknown-linux-musl"
+        ;;
+    Linux:mips*)     
+        TARGET="mips-unknown-linux-musl"
+        ;;
+    FreeBSD:x86_64*)
+        TARGET="x86_64-unknown-freebsd"
+        ;;
+    FreeBSD:amd64*)
+        TARGET="x86_64-unknown-freebsd"
+        ;;
+    FreeBSD:i686*)
+        TARGET="i686-unknown-freebsd"
+        ;;
+    Darwin:x86_64*)    
+        TARGET="x86_64-apple-darwin"
+        ;;
+    WindowsNT:x86_64*)    
+        TARGET="x86_64-pc-windows-msvc.exe"
+        ;;
+    WindowsNT:i686*)    
+        TARGET="i686-pc-windows-msvc.exe"
+        ;;
+    *)          
+        echo "Unsupported system ($(uname -a))"
+        exit 1
+        ;;
+    esac
 
-if [ -z "$TMPDIR" ]
-then
-    TMPDIR="/tmp"
-fi
-
-TEMP_PATH="$TMPDIR/tunshell"
-CLIENT_PATH="$TEMP_PATH/client"
-HEADERS_PATH="$TEMP_PATH/headers"
+    if [ ! -z "$XDG_CACHE_HOME" ]
+    then
+        TEMP_PATH="$XDG_CACHE_HOME"
+    elif [ ! -z "$TMPDIR" ]
+    then
+        TEMP_PATH="$TMPDIR"
+    elif [ -x "$(command -v mktemp)" ]
+    then
+        TEMP_PATH="$(mktemp -d)"
+    else
+        TEMP_PATH="/tmp"
+    fi
 
-mkdir -p $TEMP_PATH
+    TEMP_PATH="$TEMP_PATH/tunshell"
+    CLIENT_PATH="$TEMP_PATH/client"
+    HEADERS_PATH="$TEMP_PATH/headers"
 
+    mkdir -p $TEMP_PATH
 
-if [ -x "$(command -v curl)" ]
-then
-    INSTALL_CLIENT=true
+    if [ ! -O $TEMP_PATH -a -z "$TUNSHELL_INSECURE_EXEC" ];
+    then
+        echo "Temp path $TEMP_PATH is not owned by current user"
+        echo "Run again with TUNSHELL_INSECURE_EXEC=1 to ignore this warning" 
+        exit 1
```

Comment:
IDK what operating systems allow giving files to other users but this check still seems likely to thwart attacks at least on most Linux systems.

Reply:
Running as root or facls? Would be an edge case but I was able reproduce by running as the current user and then re-running as root. So worth having the override option available.

Reply:
I mean where an attacker does this:

```
mkdir /tmp/tunshell
cp evilprog /tmp/tunshell/client
chown you /tmp/tunshell/client
```

This is disallowed by default on most Linux distributions but I think it can be enabled. (IIRC it isn't intended as a security feature, it is to avoid bypassing disk quotas)

A less likely option would be to find a directory with a malicious file called `client` owned by the target user and symlinking that into the target location. 

I think the TL;DR is that we shouldn't be executing files that happen to be lying around in /tmp even if we think that they are a cache from the last time the user (or root) ran tunshell.

Reply:
See https://github.com/TimeToogo/tunshell/pull/21#discussion_r617513378, checksum verification should mitigate the attack you have described

Reply:
This is still vulnerable to to a time-of-check time-of-use vulnerability. I don't think that is really avoidable unfortunately. If you are really against trying harder to put the binary in `$HOME` or another trusted location than I think this is the best that we can do. I can't think of a way on operating systems that don't allow changing the owner of directories to other users.

But I will repeat for posterity: I think that executing binaries that happen to exist in a world-writable location is a bad idea and I would strongly advise against it no matter how many checks you think you have added.

Reply:
Agree with your sentiment. Now we will only ever used the cached binary if it's sha256 hash can be verified. 
Cannot think of anyway to prevent the TOCTOU without introducing more complexity than it is worth.
Going to get this merged in the current state, if you have any further comments feel free to open an new issue.